service/header/core_listener.go

```diff
@@ -38,93 +38,59 @@ func NewCoreListener(p2pSub *P2PSubscriber, fetcher *core.BlockFetcher, dag form
 
 // Start kicks off the CoreListener listener loop.
 func (cl *CoreListener) Start(ctx context.Context) error {
-	cl.ctx, cl.cancel = context.WithCancel(context.Background())
+	if cl.blockSub != nil {
+		return fmt.Errorf("core-listener: already started")
+	}
+
+	sub, err := cl.fetcher.SubscribeNewBlockEvent(ctx)
+	if err != nil {
+		return err
+	}
 
-	go cl.listen(ctx)
+	cl.ctx, cl.cancel = context.WithCancel(context.Background())
+	cl.blockSub = sub
+	go cl.listen()
 	return nil
 }
 
 // Stop stops the CoreListener listener loop.
-func (cl *CoreListener) Stop(context.Context) error {
+func (cl *CoreListener) Stop(ctx context.Context) error {
 	cl.cancel()
-	cl.ctx, cl.cancel = nil, nil
-	return nil
```

Comment:
`Start` and `Stop` are exactly the place where subscriptions should be started and stopped. This is what those methods are for. Also, it allows handling errors and not just logging them in the air.